### ISSUE TITLE
how does foo_midi use vsthost32.exe?

### ISSUE THREAD

**Issue by TheBlackMIDIMasta** [kode54/foo_midi#12]
Just typing 'vsthost32.exe' into the terminal doesn't do anything, and I would like to use it in my project for loading VSTis and sending messages. Are there any switches or parameters I'm missing?

**Comment by kode54** [comment]
If you'll read VSTiPlayer.(cpp|h), you'll see that it communicates via piped commands and data. There is also a vsthost64.exe for 64 bit instruments, and a simple detection mechanism that determines if an instrument is 32 or 64 bit before launching the appropriate host executable. It uses a really dumb checksum algorithm in a pathetic attempt to keep out meddlers who don't know what they're doing. :D

**Comment by TheBlackMIDIMasta** [comment]
So, how do I get it to select a VSTi and/or send MIDI events to it? I have an XG-compatible 32-bit VSTI called the YAMAHA S-YXG2006LE.

**Comment by kode54** [comment]
Observe how the MIDIPlayer and VSTiPlayer classes work. You call one function with a full path to a VSTi DLL, which returns true on success. Then you can call another function with MIDI commands or SysEx commands to send at the current moment, another call to render samples, and finally, calls to deal with the editor. Of course, the way it is now, the editor is only implemented for the first instance of the VSTi, which is actually instantiated three times for three ports.

**Comment by TheBlackMIDIMasta** [comment]
Would it be possible to use it as a synth? What I mean is: could you make a windows application in which I could just give it a vsti and it would listen on the dummy Midi through port 0 port for MIDI input? For example, with a syntax like this:

wine vsthostsynth.exe <VSTi DLL>

Here's a real-world example:

wine vsthostsynth.exe S-YXG50.dll

And it would start listening on Midi-through port 0 for midi input.
Narrowing it down a bit, I am only going to be using one VSTi, and its 32-bit, so no need for 64 bit executables. I have tried other programs such as Hermann Sieb's VSTHost, but the VSTis crackle and pop under wine, even after setting the buffer size to the maximum option, but your vsthost application thing seems to be the only working program that plays to and handles the VSTis

I wouldn't be able to do this myself because I'm quite programming illiterate and am scared of source code. I know it seems like a bit much, but can you accomplish all of the above for me?

**Comment by kode54** [comment]
For use in wine, it could work, but it would theoretically need to be alterd to serve multiple sessions via local TCP/IP instead of pipes.

**Comment by TheBlackMIDIMasta** [comment]
No, not those kinds of ports, I mean the virtual ports like the one created by bassmididrv.

**Comment by kode54** [comment]
I don't believe that wine interfaces to ALSA in such a way that it can accept MIDI input. You'd need to create the virtual MIDI port with a Linux native application first.

**Comment by TheBlackMIDIMasta** [comment]
Which I have done. I already have a dummy virtual port ready for use.

**Comment by kode54** [comment]
I have no idea how to open MIDI input ports in Windows, but it should be doable, maybe. Assuming wine even supports MIDI input.

**Comment by TheBlackMIDIMasta** [comment]
I don't want wine to create a midi port, I want it to listen on an already-existent port created by a native linux program.

**Comment by kode54** [comment]
And I have no idea if wine can even read from ALSA MIDI inputs. I'm not a wine developer, I don't know what will work with it.

**Comment by TheBlackMIDIMasta** [comment]
Trust me it will, I've seen programs that can do this, but listening on certain ports for midi data has next to nothing to do with wine. Developers do this all the time and prompt the user to download a virtual MIDI loopback cable program like MIDI-Yoke, but this is not needed on Linux as the MIDI through port 0 dummy ports exist as both an input and output port.

**Comment by kode54** [comment]
Then get to work and make it happen. I'll gladly allow the use of this code in another project, I just need to push a license into the repository. Of course, the license on the VST SDK may not be as permissive as Open Source needs.

**Comment by TheBlackMIDIMasta** [comment]
Ok, what did you compile sthost32.exe with? MinGW?

**Comment by TheBlackMIDIMasta** [comment]
Okay I will! :D

In the famous words of Shia LaBeouf: MAKE YOUR DREAMS COME TRUE!

**Comment by TheBlackMIDIMasta** [comment]
You know what? Screw the whole synth idea! >:( If I cant even play a midi file with another VST host, then there's no way in hell that its possble twrite a SYNTH! I mean listen to this sh!t!

http://www.mediafire.com/download/z64eapebhbrznsv/Shitty+VST+under+Linux.wav

That was a recording of me playing a midi with the S-YXG50 while using another VST host.

Here is a video for proof of concept, and its even worse in the video!:

https://www.youtube.com/watch?v=0pTbFJMco7A

And, like I said, I AM PROGRAMMING I̲L̲L̲I̲T̲E̲R̲A̲T̲E̲!̲ And it's exactly as it sounds. I can't code. And like I said, forget the synth idea. Now I just want a MIDI PLAYER. A standalone command line executable with a syntax like this:

vstplayer.exe [VSTi] [midi file]

I don't even need to access the interface of the vst, no vst editors or anything like that. I just nned something that can read off a midi player and send the messages to the vsti.

And like I said, I can't do this myself. The rest of my family are REALLY computer illiterate and probably couldn't tell the difference between a system running windows 95 and a system running windows 8.

**Comment by lybxlpsv** [comment]
weird, I just tried it on my Linux VPS and it doesn't have any crackling https://www.youtube.com/watch?v=j2WMkVj4DHo. you sure you have latest Wine? (64bit winearch works better here somehow.) also try use pulseaudio.
also mind if you share the vstplayer binary or source :p

**Comment by TheBlackMIDIMasta** [comment]
The crackling onlly happens if you use it a lot.

**Comment by TheBlackMIDIMasta** [comment]
Point being, kode54, can you make a standalone VST MIDI player?

**Comment by TheBlackMIDIMasta** [comment]
Hey kode54 you there?
